All pending `dacite` PRs - merged and released as `dacite2`

idanmiara opened the pull request:
Hi,

Edit: I pushed this to pypi as `dacite2==2.0.0`, feel free to test.
https://pypi.org/project/dacite2/

@konradhalas - I made this fork as I wanted to merge all the pending PRs. 
Since I can't push to pypi with the name `dacite` I renamed the pypi package to `dacite2` (module name still `dacite`).
If you want to take this from here and merge the PRs to `dacite` please do :)

I've tried to rebase all the open PRs one by one.
I tied to handle the conflicts as best as I could.
But I didn't dive deeply into each and every PR, so there might be bugs. 
Feel free to comment!

@thincal @das-intensity @radiosilence @antonagestam @michalchrobok @SzymonSwic @BurningKarl @daiwt @tgpfeiffer @yerihyo @trojkat @sinback @vvidic @channelcat
 
https://github.com/konradhalas/dacite/pull/199
https://github.com/konradhalas/dacite/pull/196
https://github.com/konradhalas/dacite/pull/191
https://github.com/konradhalas/dacite/pull/184
https://github.com/konradhalas/dacite/pull/175
https://github.com/konradhalas/dacite/pull/172
https://github.com/konradhalas/dacite/pull/164
https://github.com/konradhalas/dacite/pull/159
https://github.com/konradhalas/dacite/pull/156
https://github.com/konradhalas/dacite/pull/128
https://github.com/konradhalas/dacite/pull/122
https://github.com/konradhalas/dacite/pull/120
https://github.com/konradhalas/dacite/pull/86
https://github.com/konradhalas/dacite/pull/83
https://github.com/konradhalas/dacite/pull/76

```
git remote add antonagestam https://github.com/antonagestam/dacite.git
git remote add thincal https://github.com/thincal/dacite.git
git remote add das-intensity https://github.com/das-intensity/dacite.git
git remote add SzymonSwic https://github.com/SzymonSwic/dacite.git
git remote add BurningKarl https://github.com/BurningKarl/dacite.git
git remote add daiwt https://github.com/daiwt/dacite.git
git remote add yerihyo https://github.com/yerihyo/dacite.git
git remote add trojkat https://github.com/trojkat/dacite.git
git remote add Pickle-Robot https://github.com/Pickle-Robot/dacite.git
git remote add vvidic https://github.com/vvidic/dacite.git
git remote add channelcat https://github.com/channelcat/dacite.git
git remote add michalchrobok https://github.com/michalchrobok/dacite.git
git remote add tgpfeiffer https://github.com/tgpfeiffer/dacite.git

git fetch antonagestam
git fetch thincal
git fetch das-intensity
git fetch SzymonSwic
git fetch BurningKarl
git fetch daiwt
git fetch yerihyo
git fetch trojkat
git fetch Pickle-Robot
git fetch vvidic
git fetch channelcat
git fetch michalchrobok
git fetch tgpfeiffer

git merge antonagestam/fix/handle-pep-604-union-type
git merge thincal/feat/support-missing-fields
git merge das-intensity/master
git merge daiwt/master
git merge BurningKarl/master
git merge vvidic/patch-1
git merge yerihyo/yerihyo-20210410
git merge trojkat/master
git merge SzymonSwic/master
git merge channelcat/master
git merge michalchrobok/fix/empty-tuple-in-sequence-type
git merge antonagestam/feature/strict-type-checks
git merge Pickle-Robot/type-hooks-is-mapping
git merge tgpfeiffer/155-generic-ndarray
```

antonagestam committed:
Add stricter mypy type check flags

- Remove unused type: ignore.
- Remove asterisk import and add `__all__` with all exported names, to
  pass with the `no_explicit_reexport` mypy flag.
- Add various mypy config for developer ergonomics (prettier errors,
  printing error codes, warning about unused config, ignores and casts
  etc). Kept `warn_return_any = False` in `dacite.types` for now since there
  would have to be a lot of ignores/casts otherwise.

Commit:
Add generic hook types

vvidic committed:
Fix encoding of PKG-INFO file

Reproducible build fails if the file is encoded differently depending on the current locale settings when reading README.md.

sinback committed:
types: more general type hint for type_hooks

Also add a test to help ensure that type_hooks continues to work with a
very general Mapping implementation.

trojkat committed:
from_dict customization per data class

Commit:
.gitignore updated

tgpfeiffer committed:
do not treat generic numpy.ndarray as a generic collection
allow `is_instance` checks for arbitrary generic types
add test for `npt.NDArray` handling

daiwt committed:
fix #92. change Data from Dict to Mapping

BurningKarl committed:
Fix extract_optional for Optional of Union.

Commit:
Fix Type_hooks not applied to InitVar fields #158
test for type_hook with initVar added

michal-chrobok committed:
fix #168 empty tuple used for Sequence type, added tests
added tests for type hinting with generic collections

antonagestam committed:
Support PEP 604 unions, types.UnionType

das-intensity committed:
fixed from_dict in case of frozen dataclass with non-init field with default value

Commit:
feat: add `allow_missing_fields_as_none` option for missing fields in dict assigned with None value

idanmiara committed:
linting and fix issues

idanmiara committed:
fix some tests
mark 2 tests with `init_var_type_support`

idanmiara committed:
Drop support for Python 3.6 and Add support for Python 3.11

idanmiara committed:
test_ndarray.py - skip if numpy is not installed

idanmiara commented:
Finally all the tests pass.
Please review @thincal @das-intensity @radiosilence @antonagestam @michalchrobok @SzymonSwic @BurningKarl @daiwt @tgpfeiffer @yerihyo @trojkat @sinback @vvidic @channelcat

antonagestam commented:
@idanmiara While I'm sympathetic to this initiative, I'm not sure I can add much value in reviewing all these PRs as one in bulk. That should really happen one by one in my opinion.

I do stand by the PRs I've opened myself, and I still think they're valuable contributions.

If you want help in reviewing individual PRs, I could possibly commit some of my time to that. I do think it's unfortunate that it's come to this point with the original project, but at some point things just have to move on. Ideally we could work out a way to move forward with the blessing of @konradhalas to avoid having a split community.

idanmiara committed:
CHANGELOG.md - updated

idanmiara commented:
Hi @antonagestam 
Thanks for your inputs.

> @idanmiara While I'm sympathetic to this initiative, I'm not sure I can add much value in reviewing all these PRs as one in bulk. That should really happen one by one in my opinion.
> 
> I do stand by the PRs I've opened myself, and I still think they're valuable contributions.
> If you want help in reviewing individual PRs, I could possibly commit some of my time to that. 

Good point. I've rebased all everything and squashed each PR by itself, and solved all conflicts.
I think it should be much easier to review now individual commits. feel free to do so.

>I do think it's unfortunate that it's come to this point with the original project, but at some point things just have to move on. Ideally we could work out a way to move forward with the blessing of @konradhalas to avoid having a split community.

I've decided to publish in pypi as `dacite2` and keep the module name `dacite` so switching between the two is super easy. 
Also in case @konradhalas wants to step in and merge some of this work.
